src/os/bluestore/BlueStore.cc

```diff
@@ -10733,7 +10733,6 @@ int BlueStore::_do_remove(
   bool is_gen = !o->oid.is_no_gen();
   _do_truncate(txc, c, o, 0, is_gen ? &maybe_unshared_blobs : nullptr);
   if (o->onode.has_omap()) {
-    o->flush();
     _do_omap_clear(txc,
```

Comment:
_do_omap_clear has to iterator over the keys in order to delete them, so the flush() is needed here.  We could  move it into _do_omap_clear for better clarity, perhaps

Reply:
now it use rm_range_keys . so we only consider rm_range_keys

Reply:
(sorry, I am unfamiliar with RocksDB)

@liewegas The root cause here is because there are still chances we are manipulating the omap of the same object in concurrent transaction groups, such as pglogs?

Reply:
for example, if following osd_op rm the key which previous op add,but the previous don't submit .so db don't know key exist.

Reply:
Thanks, jianpeng.



> if following osd_op rm the key which previous op add,but the previous don't submit .so db don't know key exist.

Then **add_omap_keys** and **rm_omap_keys** should be two separate ops, right?  Generally PG should order write ops manipulating the same object, which means **rm_omap_keys** should not be handled by BlueStore until **add_omap_keys** is stable, but I guess there might be exceptions(e.g., pglogs?)

Reply:
exactly.. when i fixed this the test failure i saw was that the pg log was removed but left a single omap key behind, confusing things down the line.

the problem is that in order to implement the rm_range_keys we have to enumerate them which means we have to flush prior writes.

someday when rocksdb let's us enable the RangeDelete we can take these out.. but not until then

Reply:
it's fine to have an add followed by a delete.. just not a range delete that requires us to enumerate in order to name the keys to delete.

Reply:
@liewegas  Thanks for clarifications:-)